src/book/book.rs

```diff
@@ -152,15 +152,20 @@ pub struct Chapter {
     pub sub_items: Vec<BookItem>,
     /// The chapter's location, relative to the `SUMMARY.md` file.
     pub path: PathBuf,
+    /// An ordered list of the names of each chapter above this one, in the hierarchy.
+    pub parent_names: Vec<String>,
```

Comment:
Seeing that we need to manually store the parent names and keep them in sync makes me question whether it may be better to represent a `Book` as containing a "doubly-linked" tree of chapters, or even turn it into a graph structure. That way a `Chapter` can contain a pointer back to its parent and you can use that to walk up and down the levels of the tree. 

It would also hopefully simplify tracking section numbers and parsing. I might play around with this idea on a private branch and see how it goes.